compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs

```diff
@@ -0,0 +1,126 @@
+//! Migration code for the `expr_fragment_specifier_2024`
+//! rule.
+use tracing::debug;
+
+use rustc_ast::token::Token;
+use rustc_ast::token::TokenKind;
+use rustc_ast::tokenstream::TokenStream;
+use rustc_ast::tokenstream::TokenTree;
+use rustc_session::declare_lint;
+use rustc_session::declare_lint_pass;
+use rustc_session::lint::FutureIncompatibilityReason;
+use rustc_span::edition::Edition;
+use rustc_span::sym;
+
+use crate::lints::MacroExprFragment2024;
+use crate::EarlyLintPass;
+
+declare_lint! {
+    /// The `edition_2024_expr_fragment_specifier` lint detects the use of `expr` fragments
+    /// during migration to the 2024 edition.
+    ///
+    /// The `expr` fragment specifier will accept more expressions in the 2024 edition.
+    /// To maintain the current behavior, use the `expr_2021` fragment specifier.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,edition2021,no_run,no_compile
+    /// #![deny(edition_2024_expr_fragment_specifier)]
+    /// macro_rules! m {
+    ///   ($e:expr) => {
+    ///       $e
+    ///   }
+    /// }
+    ///
+    /// fn main() {
+    ///    m!(1);
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Rust [editions] allow the language to evolve without breaking
+    /// backwards compatibility. This lint catches code that uses new keywords
+    /// that are added to the language that are used as identifiers (such as a
+    /// variable name, function name, etc.). If you switch the compiler to a
+    /// new edition without updating the code, then it will fail to compile if
+    /// you are using a new keyword as an identifier.
+    ///
+    /// FIXME: change this: You can manually change the identifiers from `expr` to `expr2021`,
+    ///  or use a [raw identifier], for example `r#gen`, to transition to a new edition.
+    ///
+    /// This lint solves the problem automatically. It is "allow" by default
+    /// because the code is perfectly valid in older editions. The [`cargo
+    /// fix`] tool with the `--edition` flag will switch this lint to "warn"
+    /// and automatically apply the suggested fix from the compiler (which is
+    /// to use a raw identifier). This provides a completely automated way to
+    /// update old code for a new edition.
+    ///
+    /// [editions]: https://doc.rust-lang.org/edition-guide/
+    /// [raw identifier]: https://doc.rust-lang.org/reference/identifiers.html
+    /// [`cargo fix`]: https://doc.rust-lang.org/cargo/commands/cargo-fix.html
+    pub EDITION_2024_EXPR_FRAGMENT_SPECIFIER,
+    Allow,
+    "The `expr` fragment specifier will accept more expressions in the 2024 edition. \
+    To keep the existing behavior, use the `expr_2021` fragment specifier.",
+    @future_incompatible = FutureIncompatibleInfo {
+        reason: FutureIncompatibilityReason::EditionSemanticsChange(Edition::Edition2024),
+        reference: "issue #123742 <https://github.com/rust-lang/rust/issues/123742>",
```

Comment:
I wonder if it's better to link to RFC 3531?

Reply:
Linking to the page in the edition guide would be the best.

Reply:
We do not have a real edition guide for this @traviscross, or are you saying just to point the user to the edition guide?

At the moment I added the RFC that @eholk, but we can add also the edition guide if you like.

Trying to address in the commit d665cd868a0d6ecfec0d07acbe8de10f3b879772****

Reply:
I agree that the edition guide would be best but first we have to write that section of the edition guide :)

Why don't we link the RFC for now and update it with a link to the edition guide once we have it?

Reply:
On the edition guide, if you want to put up PR just for a stub page (that itself for now just links elsewhere), so that can be linked here, we can merge that.  E.g., [here's an existing stub](https://doc.rust-lang.org/stable/edition-guide/rust-2024/rustfmt-overflow-delimited-expr.html).  (I'd put it up myself but then I couldn't merge it.)

Reply:
Thanks, I think makes sense, we can do it

Reply:
OK added the migration guide, let me know if you want change the wording before I will squash the commtis

Reply:
See: https://github.com/rust-lang/rust/pull/125627#discussion_r1630089274

Reply:
Linking the RFC for now, then then we arrive at the doc step we rework this section, putting a note on my todo list